### PR TITLE
 RuntimeError: NYI: Named tensors are not supported with the tracer

### LINKED ISSUE #48054: RuntimeError: NYI: Named tensors are not supported with the tracer
## 🐛 Bug

While running the following script, I got the error: "RuntimeError: NYI: Named tensors are not supported with the tracer" when the input is a scalar. I guess `x[x > 0]` does not support scalar `x`? 

## To Reproduce

```
import torch

def f(x):
    return x[x > 0]

jf = torch.jit.trace(f, torch.tensor(2., device="cuda"))
```
or
```
import torch

def f(x):
    return torch.distributions.HalfCauchy(x).log_prob(x)

jf = torch.jit.trace(f, torch.tensor(2., device="cuda"))
```

## Expected behavior

No error

## Environment

 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version: 11
 - GPU models and configuration:
 - Any other relevant information:

cc @gmagogsfm @fritzo @neerajprad

### PR DESCRIPTION
**Pull request by warrenburch**
Repro 1:

```python
import torch

def f(x):
    return x[x > 0]

jf = torch.jit.trace(f, torch.tensor(2., device="cuda"))
```
Error:

```bash
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/pytorch/torch/jit/_trace.py", line 874, in trace
    traced = torch._C._create_function_from_trace(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<stdin>", line 2, in f
RuntimeError: NYI: Named tensors are not supported with the tracer
```

Repro2:

```python
import torch
import torch.nn.functional as F
from torch import nn
import copy

class Net(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, inputs):
        x = copy.deepcopy(inputs) # RuntimeError: NYI: Named tensors are not supported with the tracer
        x = F.relu(x)
        return x

model = Net()
images = torch.randn(8, 28, 28)
torch.jit.trace(model, images)
```

Error 2:

```bash
Traceback (most recent call last):
  File "/opt/pytorch/test_deepcopy.py", line 18, in <module>
  File "/opt/pytorch/torch/jit/_trace.py", line 806, in trace
    return trace_module(
           ^^^^^^^^^^^^^
  File "/opt/pytorch/torch/jit/_trace.py", line 1074, in trace_module
    module._c._create_method_from_trace(
  File "/opt/pytorch/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1501, in _slow_forward
    result = self.forward(*input, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/test_deepcopy.py", line 12, in forward
    x = F.relu(x)
        ^^^^^^^^^^
  File "/opt/conda/envs/ptca/lib/python3.11/copy.py", line 153, in deepcopy
    y = copier(memo)
        ^^^^^^^^^^^^
  File "/opt/pytorch/torch/_tensor.py", line 122, in __deepcopy__
    new_storage = self._typed_storage()._deepcopy(memo)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/storage.py", line 847, in _deepcopy
    return self._new_wrapped_storage(copy.deepcopy(self._untyped_storage, memo))
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/ptca/lib/python3.11/copy.py", line 153, in deepcopy
    y = copier(memo)
        ^^^^^^^^^^^^
  File "/opt/pytorch/torch/storage.py", line 112, in __deepcopy__
    new_storage = self.clone()
                  ^^^^^^^^^^^^
  File "/opt/pytorch/torch/storage.py", line 126, in clone
    return type(self)(self.nbytes(), device=self.device).copy_(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: NYI: Named tensors are not supported with the tracer
```

---- 
 #48054 RuntimeError: NYI: Named tensors are not supported with the tracer
 #49538 jit tracer doesn't work with unflatten layer
 #31591 when i try to export a pytorch model to ONNX, got RuntimeError: output of traced region did not have observable data dependence with trace inputs; this probably indicates your program cannot be understood by the tracer.
   - This bug was closed but exists. Multiple comments on it still showing error. This is addressed

Likely fixes the following issues (but untested)

 #63297 Named tensor in tracer
 #2323 [Bug] torch.onnx.errors.UnsupportedOperatorError when convert mask2former to onnx

Fix zero dimensioned tensors when used with jit.trace They are currently assigned an empty set for names {} this is not the same as "no name" so jit.trace bails with
  "NYI: Named tensors are not supported with the tracer"
This happens when I am trying to save a non-trivial model as onnx but the simplest repro I have seen is 48054 above which has been added as test/jit/test_zero_dim_tensor_trace.py

Test plan:
  New unit test added
  Broken scenarios tested locally
  CI

Fixes #48054
@pytorchbot label "topic: not user facing"

### CODE REVIEW

**Comment by warrenburch** [comment]
I completed the attestation a few hours ago but the merging still appears blocked?

File: test/test_zero_dim_tensor_trace.py
Comment (zou3519) [inline test/test_zero_dim_tensor_trace.py:12]: Please move these tests to test/test_jit.py

**Review by zou3519 — COMMENTED** [review]
code LGTM, but please put the tests into test_jit.py instead of creating a new test file.

**Comment by some16** [comment]
Any chance this could get fixed? I'm having the issue [#48054](https://github.com/pytorch/pytorch/issues/48054)

**Comment by warrenburch** [comment]
I have submitted the PR and it is waiting on review. I dont know how
long it takes to get reviewed.
In the meantime you could just try pulling my branch (
https://github.com/warrenburch/pytorch/tree/user/warrenb/FixZeroDimensionTensorsInTracer)
and using that, you would have to rebuild pytorch yourself though which if
you havent done it before can be a little involved.
Cheers
Warren

**Review by zou3519 — DISMISSED** [review]
code LGTM, but you will need to fix the CLA issues before merging this.

**Comment by warrenburch** [comment]
Hi Richard (@zou3519), I have tried changing my name to match and have filled out the individual CLA several times. I am not sure what I need to do to get this to go through. Any help greatly appreciated! 
Cheers
Warren

**Comment by zou3519** [comment]
can you squash all the commits on your branch and then force-push to it?

**Comment by warrenburch** [comment]
@zou3519 Thanks Richard, I have squashed and pushed. I have filled out the CLA form (several times) but it doesnt seem to like it. I'm at a loss at this point how to proceed.
Cheers
Warren

**Comment by zou3519** [comment]
@warrenburch  can you try opening a new PR and seeing what happens?

**Comment by thiagocrepaldi** [comment]
Reopening as this issue is still relevant

**Comment by thiagocrepaldi** [comment]
Closing and pushing a new PR for this

**Comment by warrenburch** [comment]
And the fix is still good. I had so many problems getting this accepted because of process I just gave up when I moved on to another project. It’s be great if someone who is able to make the repo changes to this up-WarrenOn Jan 26, 2024, at 9:35 AM, Thiago Crepaldi ***@***.***> wrote:﻿
Closing and pushing a new PR for this

**Comment by warrenburch** [comment]
Excuse the autocorrect mess in the mail (on a tiny screen). I think the intention is clear-WarrenOn Jan 26, 2024, at 9:45 AM, Warren Burch ***@***.***> wrote:﻿And the fix is still good. I had so many problems getting this accepted because of process I just gave up when I moved on to another project. It’s be great if someone who is able to make the repo changes to this up-WarrenOn Jan 26, 2024, at 9:35 AM, Thiago Crepaldi ***@***.***> wrote:﻿
Closing and pushing a new PR for this

**Comment by warrenburch** [comment]
Thanks Thiago!-WarrenOn Jan 26, 2024, at 9:35 AM, Thiago Crepaldi ***@***.***> wrote:﻿
Closing and pushing a new PR for this

**Pull request closed without merging**
